### PR TITLE
feat: add match URL parsing and validation

### LINKED ISSUE #19: Add basic URL parsing and validation layer
## Summary
Separate URL input handling from note generation so the MVP can validate, normalize, and reject bad input cleanly.

## Why
Even before any external provider work exists, the plugin should not treat raw user input as trusted. A small parsing/validation layer keeps the command flow cleaner and makes later provider work easier to isolate.

## Scope
- Parse and normalize user-provided URLs.
- Detect valid versus invalid URLs.
- Detect supported versus unsupported inputs if needed for the MVP.
- Provide useful error messages for invalid or unsupported inputs.
- Return a structured representation that downstream note-generation code can use.

## Implementation expectations
- Keep this logic separate from the command entry point.
- Keep this layer independent from provider-specific network fetching.
- Prefer a simple, explicit shape over premature abstraction.

## Definition of done
- Invalid or unsupported URLs fail gracefully without breaking the command flow.
- Downstream code receives a normalized input shape instead of raw strings.
- The parsing layer is simple enough to test independently.

### PR DESCRIPTION
**Pull request by FerdiHS**
This pull request introduces a new URL parsing and normalization workflow for match notes, which improves validation and ensures that URLs are consistently structured. The changes include a new `MatchUrl` type, a robust URL parser with error handling, and updates to the schema and code to use parsed URLs instead of raw user input.

**URL parsing and normalization:**

* Added a new `MatchUrl` type (with `sourceUrl` and `sourceHost`) to represent normalized match URLs in `src/types.ts`.
* Implemented `parseMatchUrl` in `src/services/match-url-parser.ts` to validate and normalize URLs, with detailed error handling for empty, invalid, or unsupported URLs.

**Schema and interface updates:**

* Changed `MatchNoteInput` to use a `source: MatchUrl` property instead of a raw `sourceUrl` string in `src/types.ts`.
* Updated the match note draft creation logic in `src/services/match-note-template.ts` to use the normalized `source.sourceUrl` instead of the raw input.

**Documentation:**

* Clarified in `docs/match-note-schema.md` that `source_url` should be written from the plugin’s normalized URL parser output, not raw user input.

**Related Issues:**

* Resolve #19

### CODE REVIEW

**Comment by FerdiHS** [comment]
@codex review

Please review this PR with a focus on:
- URL parsing and normalization behavior
- malformed and unsupported URL handling
- note input type changes and downstream template wiring
- regressions, edge cases, and missing validation gaps

This PR is intended to resolve #19.

File: src/services/match-url-parser.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	const trimmedInput = input.trim();
+
+	if (trimmedInput.length === 0) {
+		return createParseError('empty-url', 'Match URL cannot be empty.');
+	}
+
+	let parsedUrl: URL;
+
+	try {
+		parsedUrl = new URL(trimmedInput);
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-url-parser.ts:39]: Fixed in the latest changes. The parser now rejects raw inputs containing backslashes or control characters before passing them to `new URL(...)`, so malformed inputs that WHATWG parsing would otherwise recover from are treated as `invalid-url` instead of being silently normalized.

**Comment by FerdiHS** [comment]
@codex review

Please review this PR with a focus on:
- URL parsing and normalization behavior
- malformed and unsupported URL handling
- note input type changes and downstream template wiring
- regressions, edge cases, and missing validation gaps

This PR is intended to resolve #19.

File: src/services/match-url-parser.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	if (!startsWithSupportedWebScheme(trimmedInput)) {
+		return createParseError(
+			'invalid-url',
+			'Enter a valid match URL including http:// or https://.',
+		);
+	}
+
+	if (parsedUrl.username.length > 0 || parsedUrl.password.length > 0) {
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-url-parser.ts]: Fixed in the latest changes. The parser now rejects any raw `@` in the URL authority, so empty-userinfo forms like `https://@example.com/match` and `https://:@example.com/match` are treated as unsupported auth instead of being silently normalized and accepted.

**Comment by FerdiHS** [comment]
@codex review

Please review this PR with a focus on:
- URL parsing and normalization behavior
- malformed and unsupported URL handling
- note input type changes and downstream template wiring
- regressions, edge cases, and missing validation gaps

This PR is intended to resolve #19.

File: src/services/match-url-parser.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	for (const character of input) {
+		if (character === '\\') {
+			return true;
+		}
+
+		const codePoint = character.codePointAt(0);
+
+		if (
+			codePoint !== undefined &&
+			((codePoint >= 0x00 && codePoint <= 0x1f) || codePoint === 0x7f)
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-url-parser.ts:105]: Fixed in the latest changes. The parser now rejects raw whitespace inside the URL after trimming, so malformed inputs like `https://example.com/match 123` are treated as invalid instead of being silently normalized.

**Comment by FerdiHS** [comment]
@codex review

Please review this PR with a focus on:
- URL parsing and normalization behavior
- malformed and unsupported URL handling
- note input type changes and downstream template wiring
- regressions, edge cases, and missing validation gaps

This PR is intended to resolve #19.

**Comment by FerdiHS** [comment]
@codex review

Please do one final pass on this PR and confirm whether it is ready and safe to merge.

Focus on:
- correctness and regressions
- URL parsing and validation edge cases
- note input/template integration
- test coverage for the parser changes
- any remaining bugs, typos, or logic issues

If you do not find any issues, please explicitly confirm that this PR looks ready and okay to merge.

File: src/services/match-url-parser.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+	  }
+	| {
+			ok: false;
+			error: {
+				code: MatchUrlParseErrorCode;
+				message: string;
+			};
+	  };
+
+export function parseMatchUrl(input: string): MatchUrlParseResult {
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply src/services/match-url-parser.ts:22]: Fixed in the latest changes. I added focused regression coverage for `parseMatchUrl`, including valid URL normalization, unsupported schemes, malformed whitespace and backslashes, embedded auth in the authority, allowed `@` outside the authority, and already-encoded spaces. I also added a repo `npm run test` entrypoint and the minimal lint/config wiring needed for the new test file.

**Comment by FerdiHS** [comment]
@codex review

Please do one final pass on this PR and confirm whether it is ready and safe to merge.

Focus on:
- correctness and regressions
- URL parsing and validation edge cases
- parser regression test coverage
- note input/template integration
- any remaining bugs, typos, or logic issues

If you do not find any issues, please explicitly confirm that this PR looks ready and okay to merge.

**Comment by FerdiHS** [comment]
@codex review

Please do one final pass on this PR and confirm whether it is ready and safe to merge.

Focus on:
- correctness and regressions
- URL parsing and validation edge cases
- parser regression test coverage
- note input/template integration
- any remaining bugs, typos, or logic issues

If you do not find any issues, please explicitly confirm that this PR looks ready and okay to merge.

File: package.json [hunk 5]
```diff
@@ -7,6 +7,7 @@
 	"scripts": {
 		"dev": "node esbuild.config.mjs",
 		"build": "tsc -noEmit -skipLibCheck && node esbuild.config.mjs production",
+		"test": "node --import jiti/register --test src/services/match-url-parser.test.ts",
```
[the remark this answers is not in the document]
Comment (FerdiHS) [reply package.json:10]: Fixed in the latest changes. The parser regression suite is now part of automated checks in both CI and the local pre-push gate, so URL parsing regressions cannot merge green without running `npm run test`. I also updated the contributor docs to reflect the new pre-push behavior.

**Comment by FerdiHS** [comment]
@codex review

Please do one final pass on this PR and confirm whether it is ready and safe to merge.

Focus on:
- correctness and regressions
- URL parsing and validation edge cases
- parser regression test coverage
- note input/template integration
- any remaining bugs, typos, or logic issues

If you do not find any issues, please explicitly confirm that this PR looks ready and okay to merge.

**Comment by FerdiHS** [comment]
@codex review

Please do one final pass on this PR and confirm whether it is ready and safe to merge.

Focus on:
- correctness and regressions
- edge cases and validation gaps
- parser/test coverage
- CI and pre-push workflow changes
- any remaining bugs, typos, or logic issues

If you do not find any issues, please explicitly confirm that this PR looks ready and okay to merge.

**Comment by FerdiHS** [comment]
@codex review

Please do one final pass on this PR and confirm whether it is ready and safe to merge.

Focus on:
- correctness and regressions
- edge cases and validation gaps
- parser/test coverage
- CI and pre-push workflow changes
- any remaining bugs, typos, or logic issues

If you do not find any issues, please explicitly confirm that this PR looks ready and okay to merge.
